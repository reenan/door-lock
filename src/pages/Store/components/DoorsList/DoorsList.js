import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Message } from 'semantic-ui-react'

import DoorContainer from './components/DoorContainer'

import './DoorsList.scss'
class DoorsList extends Component {
  render() {
    const {
      roles,
      doors,
      employees,
      selectedEmployee,
    } = this.props

    const doorsKeys = Object.keys(doors)
    const employee = employees[selectedEmployee]

    const roleName = (
      employee && roles[employee.role] ?
        roles[employee.role].name : 'No role defined'
    )

    return (
      <div>
        {
          doorsKeys.length > 0 ?
            <div>
              <p>
                {
                  employee ?
                    `You have selected ${employee.name} (${roleName}),
                      and these are his/hers permissions:` :
                    'Please, select an employee to continue.'
                }
              </p>

              {
                employee ? (
                  <div>
                    <p>
                      You can click on the doors to open them.
                    </p>
                    <div className='doors'>
                      {
                        doorsKeys.map((doorID) => (
                          <DoorContainer
                            key={doorID}
                            doorID={doorID}
                            selectedEmployee={selectedEmployee}
                          />
                        ))
                      }
                    </div>
                  </div>
                ) : null
              }
            </div> :
            <Message warning>
              <Message.Header>
                Please, register at least one door before continuing.
              </Message.Header>
            </Message>
        }
      </div>
    )
  }
}

DoorsList.propTypes = {
  roles: PropTypes.object.isRequired,
  doors: PropTypes.object.isRequired,
  employees: PropTypes.object.isRequired,
  selectedEmployee: PropTypes.string,
}

DoorsList.defaultProps = {
  selectedEmployee: null
}

export default DoorsList
