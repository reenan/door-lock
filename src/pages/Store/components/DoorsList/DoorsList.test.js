import React from 'react'
import DoorsList from './DoorsList'

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

const mockStore = configureStore()

import renderer from 'react-test-renderer'

it('renders without crashing', () => {
  const store = mockStore({
    store: {
      roles: {},
      doors:{},
      employees: {},
    }
  })

  const component = renderer.create(
      <Provider store={store}>
        <DoorsList
          employees={{}}
          selectedEmployee={null}
          doors={{}}
          roles={{}}
        />
      </Provider>
  )

  expect(component).toMatchSnapshot()
})

it('renders without crashing with no employee selected', () => {
  const store = mockStore({
    store: {
      roles: {},
      doors:{ mockDoor: { name: 'Mock Door' } },
      employees: { employee: { name: 'Mock Employee', role: 'mockRole' } }
    }
  })

  const component = renderer.create(
      <Provider store={store}>
        <DoorsList
          employees={{ employee: { name: 'Mock Employee', role: 'mockRole' } }}
          selectedEmployee={null}
          doors={{ mockDoor: { name: 'Mock Door' } }}
          roles={{}}
        />
      </Provider>
  )

  expect(component).toMatchSnapshot()
})

it('renders without crashing with values', () => {
  const store = mockStore({
    store: {
      roles: {},
      doors:{ mockDoor: { name: 'Mock Door' } },
      employees: { employee: { name: 'Mock Employee', role: 'mockRole' } }
    }
  })

  const component = renderer.create(
      <Provider store={store}>
        <DoorsList
          employees={{ employee: { name: 'Mock Employee', role: 'mockRole' } }}
          selectedEmployee='employee'
          doors={{ mockDoor: { name: 'Mock Door' } }}
          roles={{ mockRole: { permissions: { mockDoor: false } } }}
        />
      </Provider>
  )

  expect(component).toMatchSnapshot()
})

it('renders without crashing with values and permission', () => {
  const store = mockStore({
    store: {
      roles: { mockRole: { name: 'Mock Role', permissions: { mockDoor: true } } },
      doors:{ mockDoor: { name: 'Mock Door' } },
      employees: { employee: { name: 'Mock Employee', role: 'mockRole' } }
    }
  })

  const component = renderer.create(
      <Provider store={store}>
        <DoorsList
          employees={{ employee: { name: 'mock', role: 'mockRole' } }}
          selectedEmployee='employee'
          doors={{ mockDoor: { name: 'Mock Door' } }}
          roles={{ mockRole: { name: 'Mock Role', permissions: { mockDoor: true } } }}
        />
      </Provider>
  )

  expect(component).toMatchSnapshot()
})
