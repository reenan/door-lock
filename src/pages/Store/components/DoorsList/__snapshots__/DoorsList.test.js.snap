// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    className="ui warning message"
  >
    <div
      className="header"
    >
      Please, register at least one door before continuing.
    </div>
  </div>
</div>
`;

exports[`renders without crashing with no employee selected 1`] = `
<div>
  <div>
    <p>
      Please, select an employee to continue.
    </p>
  </div>
</div>
`;

exports[`renders without crashing with values 1`] = `
<div>
  <div>
    <p>
      You have selected Mock Employee (undefined),
                      and these are his/hers permissions:
    </p>
    <div>
      <p>
        You can click on the doors to open them.
      </p>
      <div
        className="doors"
      >
        <div
          className="door"
          onClick={[Function]}
        >
          <p>
            <span
              className="warn red"
            >
              Not allowed to open: 
            </span>
            <span>
              Mock Door
            </span>
          </p>
          <div
            className="icons"
          >
            <i
              aria-hidden="true"
              className="massive icon custom-icon door-closed"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="massive icon custom-icon door-open"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="unlock large icon"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="lock large icon"
              onClick={[Function]}
            />
          </div>
          <div
            className="ui dimmer"
            onClick={[Function]}
          >
            <div
              className="content"
            >
              <div
                className="ui loader"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders without crashing with values and permission 1`] = `
<div>
  <div>
    <p>
      You have selected mock (Mock Role),
                      and these are his/hers permissions:
    </p>
    <div>
      <p>
        You can click on the doors to open them.
      </p>
      <div
        className="doors"
      >
        <div
          className="door"
          onClick={[Function]}
        >
          <p>
            <span
              className="warn green"
            >
              Allowed to open: 
            </span>
            <span>
              Mock Door
            </span>
          </p>
          <div
            className="icons"
          >
            <i
              aria-hidden="true"
              className="massive icon custom-icon door-closed"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="massive icon custom-icon door-open"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="unlock large icon"
              onClick={[Function]}
            />
            <i
              aria-hidden="true"
              className="lock large icon"
              onClick={[Function]}
            />
          </div>
          <div
            className="ui dimmer"
            onClick={[Function]}
          >
            <div
              className="content"
            >
              <div
                className="ui loader"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
